### ISSUE TITLE
Fix handling of object storage

### ISSUE THREAD

**Issue by blink1073** [metabolite-atlas/metatlas#16]
Need to detect all changes when storing.

We need to check for objects which contain the root object.

We need to detect when the items in a list of objects changes.

**Comment by blink1073** [comment]
@benbowen, how about this model? You can `store()` a new version of your own object, or `clone()` it, creating a new unique id.  If you do not own the object, you can only `clone()` it.  This way, we do not have to ripple changes through the database when one object is updated.  

We use the following fields:
unique_id, version, prev_uid, prev_version
You can then trace back through your current versions, back through prev uid and versions to the original data source, if so desired.

When you have a reference to another unique id, you always get the latest version of that object back in a query.

**Comment by benbowen** [comment]
Started capture of these ideas in a flow diagram:
https://www.draw.io/#G0BweRoDuGcsLBN1hBOXRLRlpXT2c

**Comment by blink1073** [comment]
While trying to implement this, I realized there is a better way.
We stick with unique_id and previous_uid.
When you save, update the uid of previous entry in the table before creating a new one that points to it as the previous uid.
When you clone, you create a new unique id.
When you retrieve, by default you only get stubs of the sub-items, but you can pass in `recursive=True` to build the full model (which would take a lot longer).

**Comment by benbowen** [comment]
You are probably already doing this, but one thing to double check is to
make sure the new is actually changed from the old prior to save.  This
comparison function would be a nice convenience func for the user too.  See
whats different between two objects.

Just so I understand:
*"Save" will Create a new row in a table that points to previous.
*Previous object points down to all previous objects.
*Previous will now have a second UID that points to "branch"

So basically, all objects will have the ability to point to both previous
and future.

Is this correct?

**Comment by blink1073** [comment]
Yes, I had an internal check to make sure it actually changed before creating a new version.  Adding a diff is a feasibly addition.

Objects only point to the past.  The difference between this and what I had originally written is that the unique id of  head of the linked list remains constant during an update.  We are effectively stuffing a new link prior to the head whenever we make an update.

When you clone(), you are creating a new branch on the tree.
